### PR TITLE
Split prng_keys with JAX.

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
JAX RNG keys should not be reused (in production) according to the JAX documentation. The current usage only split the keys at the lowest level (`_sample_state_jax`). This could cause unwanted correlations. For example, we reuse the device key for each tape in a batch of tapes in `default.qubit`. 
 
**Description of the Change:**
Add functions and methods to split RNG keys. Split RNG keys to avoid reusing keys.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
#5337 

[sc-61435]

### CODE REVIEW

File: pennylane/devices/qubit/simulate.py [hunk 1]
```diff
@@ -248,11 +248,14 @@ def simulate(
     """
     has_mcm = any(isinstance(op, MidMeasureMP) for op in circuit.operations)
     if circuit.shots and has_mcm:
-        return simulate_one_shot_native_mcm(circuit, rng, prng_key, debugger, interface)
+        return simulate_one_shot_native_mcm(
+            circuit, rng, prng_key=prng_key, debugger=debugger, interface=interface
+        )
+    _, key = jax_random_split(prng_key)
```
Comment (vincentmr) [inline pennylane/devices/qubit/simulate.py:254]: If we pass rng/key to `get_final_state`, we could use the first key.

File: pennylane/devices/default_qubit.py [hunk 2]
```diff
@@ -379,6 +380,21 @@ def name(self):
...
 
+    @property
+    def prng_key(self):
+        """Get a new key with ``jax.random.split``."""
+        self._prng_key, key = jax_random_split(self._prng_key)
+        return key
+
+    def get_prng_key(self, num: int = 1):
+        """Get ``num`` new keys with ``jax.random.split``."""
+        self._prng_key, *keys = jax_random_split(self._prng_key, num=num + 1)
```
Comment (dime10) [inline pennylane/devices/default_qubit.py:391]: Why the num+1?

File: pennylane/devices/default_qubit.py [hunk 3]
```diff
@@ -379,6 +380,21 @@ def name(self):
         """The name of the device."""
         return "default.qubit"
 
+    @property
+    def prng_key(self):
+        """Get a new key with ``jax.random.split``."""
+        self._prng_key, key = jax_random_split(self._prng_key)
+        return key
+
+    def get_prng_key(self, num: int = 1):
```
Comment (dime10) [inline pennylane/devices/default_qubit.py:389]: Does the first function make sense? I think it might be unexpected that a "property" changes upon each call.

We might also want to make the second function name plural to highlight the distinction with the first.

File: pennylane/devices/qubit/sampling.py [hunk 4]
```diff
@@ -373,7 +386,11 @@ def _sum_for_single_shot(s):
         )
         return sum(c * res for c, res in zip(mp.obs.terms()[0], results))
 
-    unsqueezed_results = tuple(_sum_for_single_shot(type(shots)(s)) for s in shots)
+    keys = jax_random_split(prng_key, num=shots.num_copies)
+    print(keys)
```
Comment (dime10) [inline pennylane/devices/qubit/sampling.py:390]: This should probably be removed I'm assuming

File: pennylane/devices/default_qubit.py (hunk 2 above)
Comment (vincentmr) [reply pennylane/devices/default_qubit.py:391]: This is because whenever requesting a key, we must generate at least an extra key which is used to spawn another tree of keys; the first key is used to overwrite `_prng_key` for that purpose. Otherwise there is a risk of reusing any of the keys. This is done under the hood to prevent users from forgetting to update the "seed".

Comment (dime10) [reply pennylane/devices/default_qubit.py:385]: Nice 👍

**Review by mudit2812 — COMMENTED** [review]
Looks great! I left a couple of comments, but overall seems pretty close to approval ready

File: pennylane/devices/default_qubit.py [hunk 5]
```diff
@@ -582,7 +595,7 @@ def execute(
                 _wrap_simulate,
                 vanilla_circuits,
                 seeds,
-                [self._prng_key] * len(vanilla_circuits),
+                self.get_prng_keys(num=len(vanilla_circuits)),
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/default_qubit.py:598]: The goal of this PR is each call to `simulate` does not uses the same PRNG key. In the current setup however, multithreaded execution will not use the same keys as it branches keys with an n-ary tree while the sequential execution branches with a binary tree, and these do not yield the same key sequences in JAX. But do we want to enforce the same key sequence between sequential and multithreading?

[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/devices/default_qubit.py:602]: This can be achieved with the changes in [8546dc3](https://github.com/PennyLaneAI/pennylane/pull/5515/commits/8546dc3e619fb8103a10182a73d139022fe9ed33).

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Can you add documentation about how the PRNG keys are used to the `DefaultQubit` docstring and `DefaultQubit.execute` docstring? Probably in a "Usage details" section.

**Comment by albi3ro** [comment]
> Can you add documentation about how the PRNG keys are used to the `DefaultQubit` docstring and `DefaultQubit.execute` docstring? Probably in a "Usage details" section. Something along the lines of
> 
> ```
> If a ``jax.random.PRNGKey`` is provided, it will be used by the device when executing circuits
> with finite shots. When executing ``n`` circuits, the ``PRNGkey`` is split into ``n`` new keys
> using ``jax.random.split`` to better simulate non-deterministic behaviour.
> 
> When simulating native mid-circuit measurements with ``s`` shots, the ``PRNGKey`` is also
> split into ``s`` new keys. If the splitting is not performed, all samples will be either ``0`` or
> ``1``, effectively eliminating stochastic behaviour when executing a circuit.
> ```

I would say that information belongs in a source code comment instead, as it's an implementation detail consumers shouldn't need to know anything about.  This information mostly needs to be preserved for the next person editing this code, to make sure we don't remove the stochasticity again. But I don't think it needs to be in a docstring.

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Sorry for not pointing this out earlier, but can you add a test to verify that the results stay the same for separate calls to `DefaultQubit.execute` since we reset the prng key?

File: pennylane/devices/default_qubit.py [hunk 6]
```diff
@@ -634,11 +673,11 @@ def execute_and_compute_derivatives(
                         _adjoint_jac_wrapper,
                         vanilla_circuits,
                         seeds,
-                        [self._prng_key] * len(vanilla_circuits),
+                        self.get_prng_keys(num=len(vanilla_circuits)),
```
Comment (albi3ro) [inline pennylane/devices/default_qubit.py:676]: Do we even need to be sending seeds when calculating derivatives? adjoint is analytic only, so we aren't going to be doing any sampling 🤔

**Review by albi3ro — COMMENTED** [review]
Might be useful to add a test along the lines of:
```
key = jax.random.key(152)

dev = qml.device('default.qubit', seed=key)

tape= qml.tape.QuantumScript([qml.Hadamard(0)], [qml.sample(wires=0)], shots=100)
sample1, sample2 = dev.execute((tape, tape))
assert not qml.math.allclose(sample1, sample2)
` ``

Edit: both with and without `max_workers`.

File: pennylane/devices/qubit/sampling.py [hunk 7]
```diff
@@ -269,6 +280,7 @@ def _process_single_shot(samples):
         return tuple(processed)
 
     try:
+        prng_key, _ = jax_random_split(prng_key)
```
Comment (albi3ro) [inline pennylane/devices/qubit/sampling.py:283]: If we split right before calling this in `measure_with_samples`, do we need to split the key again?

Comment (vincentmr) [reply pennylane/devices/qubit/sampling.py:283]: You're correct, it is not necessary here. I think we could follow either of two extremes: split only when necessary, or split prior to passing to any function. The former is cleaner. The latter is a bit cumbersome, but then we need not think too much about which key ends up where. If we go with the former I could remove a couple other split in the current PR.

**Review by albi3ro — APPROVED** [review]
Looks good! 👍

File: pennylane/devices/default_qubit.py (hunk 6 above)
Comment (vincentmr) [reply pennylane/devices/default_qubit.py]: Good point. `rng`s, `seed`s removed.

**Review by mudit2812 — APPROVED** [review]
Looks great! One final comment
